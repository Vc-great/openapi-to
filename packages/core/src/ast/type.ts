import type { JSDocTagStructure } from "ts-morph";
import type {
  AmbientableNodeStructure,
  ExportableNodeStructure,
  JSDocableNodeStructure,
  Structure,
} from "ts-morph";
import type {
  BindingNamedNodeStructure,
  ExclamationTokenableNodeStructure,
  KindedStructure,
  OptionalKind,
  StructureKind,
  TypedNodeStructure,
  VariableDeclarationKind,
  WriterFunction,
} from "ts-morph";

export interface VariableStatementStructureOfObject
  extends Structure,
    VariableStatementSpecificStructure,
    JSDocableNodeStructure,
    AmbientableNodeStructure,
    ExportableNodeStructure {}

interface VariableStatementSpecificStructure
  extends KindedStructure<StructureKind.VariableStatement> {
  declarationKind?: VariableDeclarationKind;
  declarations: OptionalKind<VariableDeclarationStructure>[];
}

export interface VariableDeclarationStructure
  extends Structure,
    VariableDeclarationSpecificStructure,
    BindingNamedNodeStructure,
    InitializerExpressionableNodeStructure,
    TypedNodeStructure,
    ExclamationTokenableNodeStructure {}

interface VariableDeclarationSpecificStructure
  extends KindedStructure<StructureKind.VariableDeclaration> {}

export interface InitializerExpressionableNodeStructure {
  initializer?: string | WriterFunction | ObjectStructure[];
}

export interface ObjectStructure {
  key: string;
  value: string;
  docs: JSDocStructure[];
}

export interface JSDocStructure {
  /**
   * The description of the JS doc.
   * @remarks To force this to be multi-line, add a newline to the front of the string.
   */
  description?: string;
  /** JS doc tags (ex. `&#64;param value - Some description.`). */
  tags?: OptionalKind<JSDocTagStructure>[];
}

export type JSDoc = JSDocStructure[];
