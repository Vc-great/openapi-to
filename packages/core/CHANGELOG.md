# @openapi-to/core

## 3.0.1

### Patch Changes

- fix bug

## 3.0.0

### Major Changes

- fix bug
- fix: bug
- update
- refactor

## 3.0.0-alpha.5

### Major Changes

- fix: bug

## 3.0.0-alpha.4

### Major Changes

- fix bug

## 3.0.0-alpha.3

### Major Changes

- refactor

## 3.0.0-alpha.2

### Major Changes

- fix

## 2.4.0-alpha.1

### Patch Changes

- fix bug

## 2.4.0-alpha.0

### Minor Changes

- update all

## 2.3.0

### Minor Changes

- generate swr

## 2.2.0

### Minor Changes

- fix bug

## 2.1.0

### Minor Changes

- optimize request type zod

## 2.0.0

### Major Changes

- [`e519466`](https://github.com/Vc-great/openapi-to/commit/e5194667c7416e817a498d592c357a7ae9c05f22) Thanks [@Vc-great](https://github.com/Vc-great)! - v2

- [`a39a900`](https://github.com/Vc-great/openapi-to/commit/a39a9002dda434d8a65768f55c69875ed8ad1eea) - fix bug

- 2.0

- [`367be25`](https://github.com/Vc-great/openapi-to/commit/367be252aa434487c09c4566e77792839867b509) - fix bug

- [`7fd9e58`](https://github.com/Vc-great/openapi-to/commit/7fd9e58417ef5563dedf945fbb030b70b8b09bd6) - rc

## 2.0.0-rc.5

### Major Changes

- rc

## 2.0.0-alpha.4

### Major Changes

- fix bug

## 2.0.0-alpha.3

### Major Changes

- fix bug

## 2.0.0-alpha.2

### Major Changes

- v2
