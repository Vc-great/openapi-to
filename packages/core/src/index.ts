export * from "./defineConfig";
export * from "./errors.ts";
export { Generator } from "./Generator.ts";
export { PackageManager } from "./PackageManager.ts";
// dprint-ignore
export { SchemaGenerator } from "./SchemaGenerator.ts";
export * from "./types.ts";
/*
export default build;*/
export { AST } from "./ast/ast.ts";
export { build, requestRemoteData } from "./build.ts";
export { OpenAPI } from "./openapi/OpenAPI.ts";
export { createPlugin } from "./plugins.ts";

/*



export interface _Register {}
export type Plugins = _Register;
export type OptionsPlugins = { [K in keyof Plugins]: Plugins[K]["options"] };
* */
