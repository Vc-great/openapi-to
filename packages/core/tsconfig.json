{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "types": ["vitest/globals", "bun-types"]
  },
  "include": ["src/**/*.ts", "./package.json", "vitest.config.ts", "tsup.config.ts", "../config-ts/reset.d.ts"]
}
