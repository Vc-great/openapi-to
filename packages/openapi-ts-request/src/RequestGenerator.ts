import path from "node:path";

import { URLPath } from "@openapi-to/core/utils";

import _ from "lodash";
import { StructureKind, VariableDeclarationKind } from "ts-morph";

import type { PluginContext } from "@openapi-to/core";
import type { AST, OpenAPI } from "@openapi-to/core";
import type { OpenapiToSingleConfig } from "@openapi-to/core";
import type Oas from "oas";
import type { Operation } from "oas/operation";
import type { OpenAPIV3 } from "openapi-types";
import type {
  ImportDeclarationStructure,
  MethodDeclarationStructure,
  OptionalKind,
} from "ts-morph";
import type { PluginConfig } from "./types.ts";

type RequestGeneratorParams = {
  oas: Oas;
  openapi: OpenAPI;
  ast: AST;
  pluginConfig: PluginConfig;
  openapiToSingleConfig: OpenapiToSingleConfig;
};

type ImportStatementsOmitKind = Omit<ImportDeclarationStructure, "kind">;

export class RequestGenerator {
  private operation: Operation | undefined;
  private oas: RequestGeneratorParams["oas"];
  private readonly paramsZodSchema: string;
  private readonly openapi: RequestGeneratorParams["openapi"];
  private readonly ast: RequestGeneratorParams["ast"];
  private readonly pluginConfig: RequestGeneratorParams["pluginConfig"];
  private readonly openapiToSingleConfig: RequestGeneratorParams["openapiToSingleConfig"];

  constructor({
    oas,
    openapi,
    ast,
    pluginConfig,
    openapiToSingleConfig,
  }: RequestGeneratorParams) {
    this.oas = oas;
    this.ast = ast;
    this.pluginConfig = pluginConfig;
    this.openapiToSingleConfig = openapiToSingleConfig;
    this.openapi = openapi;
    this.paramsZodSchema = "paramsZodSchema";
  }

  get namespaceTypeName(): string {
    return this.className + "Type";
  }

  get namespaceZodName(): string {
    return this.className + "Zod";
  }

  get isCreateZodDecorator(): boolean {
    return this.pluginConfig.createZodDecorator;
  }

  build(context: PluginContext): void {
    _.mapValues(this.openapi.pathGroupByTag, (pathGroup, tag) => {
      const methodsStatements = _.map(pathGroup, ({ path, method, tag }) => {
        this.operation = this.openapi.setCurrentOperation(path, method, tag);
        return this.generatorMethod();
      });
      const filePath = path.resolve(
        context.output,
        this.lowerFirstClassName + ".ts",
      );
      return this.ast.createSourceFile(filePath, {
        statements: [
          ...this.generateImport(),
          this.generatorClass(methodsStatements),
          ...this.generatorExport(),
        ],
      });
    });
  }

  /**
   * @example
   * ```
   * const apiName = new ApiName
   * export { apiName }
   * ```
   */
  generatorExport() {
    return [
      this.ast.generateVariableStatements({
        declarationKind: VariableDeclarationKind.Const,
        declarations: [
          {
            name: this.lowerFirstClassName,
            initializer: "new" + " " + this.className,
          },
        ],
        isExported: true,
      }),
      /*      this.ast.generateExportDeclarationStatements({
        kind: StructureKind.ExportDeclaration,
        namedExports: "apiName",
      }),*/
    ];
  }

  /**
   *
   * @example
   * ```
   * #type
   * import {} './type'
   * #zod
   * //zod decorator
   * import { paramsZodSchema, responseZodSchema, zodValidate } from '@/utils/zod'
   * // ts type
   * import type { DrmsDynamicDataType } from './drmsDynamicDataZod'
   * // zod
   * import { Zod } from './drmsDynamicDataZod'
   * ```
   */
  generateImport() {
    const zodDecorator: ImportStatementsOmitKind = {
      namedImports: ["paramsZodSchema", "responseZodSchema", "zodValidate"],
      moduleSpecifier: "@/utils/zod",
    };

    const request: ImportStatementsOmitKind = {
      namedImports: ["request"],
      moduleSpecifier: "./request",
    };

    //todo model
    //requestBody response
    const typeModel: ImportStatementsOmitKind = {
      namedImports: [this.namespaceTypeName],
      moduleSpecifier: `./${this.namespaceTypeName}`,
    };

    const errorType: ImportStatementsOmitKind = {
      namedImports: ["ErrorResponse"],
      moduleSpecifier: "./model/ErrorResponse",
    };

    const zodModel: ImportStatementsOmitKind = {
      namedImports: [this.namespaceZodName],
      moduleSpecifier: `./${this.namespaceZodName}`,
    };

    const statements = _.chain([] as Array<ImportStatementsOmitKind>)
      .concat(this.isCreateZodDecorator ? [zodModel, zodDecorator] : [])
      .push(typeModel)
      .push(errorType)
      .push(request)
      .filter(Boolean)
      .value();

    return this.ast.generateImportStatements(statements);
  }

  get className() {
    return (
      _.upperFirst(
        _.camelCase(
          this.openapi &&
            this.openapi.currentTagMetadata &&
            this.openapi.currentTagMetadata.name,
        ),
      ) + "API"
    );
  }

  get lowerFirstClassName() {
    return _.lowerFirst(this.className);
  }

  generatorClass(
    methodsStatements: OptionalKind<MethodDeclarationStructure>[],
  ) {
    const statements = {
      name: this.className,
      docs: [
        {
          description: "\n",
          tags: [
            {
              tagName: "tag",
              text:
                this.openapi &&
                this.openapi.currentTagMetadata &&
                this.openapi.currentTagMetadata.name,
            },
            {
              tagName: "description",
              text:
                this.openapi &&
                this.openapi.currentTagMetadata &&
                this.openapi.currentTagMetadata.description,
            },
          ],
        },
      ],
      methods: methodsStatements,
    };
    return this.ast.generateClassStatements(statements);
  }

  /**
   * @example
   * ```
   *@zodValidate
   *@responseZodSchema(ZOD.updateResponse)
   * ```
   */
  generatorMethodDecorators() {
    return [
      {
        name: "zodValidate",
      },
      {
        name: "responseZodSchema",
        arguments: [this.getParametersName("zod", "responseName")],
      },
    ];
  }

  /**
   *
   * @param prefix
   * @param name
   */
  getParametersName(
    prefix: "type" | "zod",
    name:
      | "pathRequestName"
      | "bodyRequestName"
      | "queryRequestName"
      | "responseName",
  ) {
    const prefixType = {
      type: this.namespaceTypeName,
      zod: this.namespaceZodName,
    };
    return prefixType[prefix] + "." + _.get(this.openapi, name);
  }

  /**
   * @example
   * ```
   * query: ApiType.FindAllQueryRequest
   * body: ApiType.UpdateBodyRequest
   * id: ApiType.FindByIdPathRequest['id']
   * ```
   */
  generatorMethodParameters() {
    const queryParameters = {
      name: "query",
      type: this.getParametersName("type", "queryRequestName"),
      decorators: [
        {
          name: this.paramsZodSchema,
          arguments: [this.getParametersName("zod", "queryRequestName")],
        },
      ],
    };

    const bodyParameters = {
      name: "body",
      type: this.getParametersName("type", "bodyRequestName"),
      decorators: [
        {
          name: this.paramsZodSchema,
          arguments: [this.getParametersName("zod", "bodyRequestName")],
        },
      ],
    };

    const pathParameters = _.chain(this.openapi.parameter?.parameters)
      .filter(["in", "path"])
      .map((item: OpenAPIV3.ParameterObject) => {
        return {
          name: _.camelCase(item.name),
          type: this.getParametersName("type", "pathRequestName"),
          decorators: [
            {
              name: this.paramsZodSchema,
              arguments: [this.getParametersName("zod", "pathRequestName")],
            },
          ],
        };
      })
      .value();

    return _.chain([] as any[])
      .push(
        this.openapi.parameter?.hasQueryParameters
          ? queryParameters
          : undefined,
      )
      .push(
        this.openapi.requestBody?.hasRequestBody ? bodyParameters : undefined,
      )
      .concat(
        this.openapi.parameter?.hasPathParameters ? pathParameters : undefined,
      )
      .filter(Boolean)
      .value();
  }

  /**
   * @example
   * ```
   * export namespace tag {
   *  export type queryRequest = object
   * }
   * ```
   */
  generateNamespaceType() {
    return {
      kind: StructureKind.Module,
      docs: [{ description: this.openapi.currentTagMetadata?.description }],
      isExported: true,
      name: this.namespaceTypeName,
      statements: [
        {
          kind: StructureKind.TypeAlias,
          isExported: true,
          name: "queryRequest",
          docs: [{ description: "queryRequest" }],
          type: "object",
        },
      ],
    };
  }

  /**
   * @example
   * ```
   * Promise<[ApiType.ErrorResponse, ApiType.FindByIdResponse]>
   * ```
   */
  generatorReturnType() {
    const resultType = this.namespaceTypeName + "." + this.openapi.responseName;
    return `Promise<[ErrorResponse,${resultType}]>`;
  }

  /**
   * @example
   * ```
   *  @summary 查询事项目录详情
   *  @description 详细描述
   * ```
   */
  generatorMethodDocs() {
    return [
      {
        description: "\n",
        tags: [
          {
            tagName: "summary",
            text: this.operation?.getSummary(),
          },
          {
            tagName: "description",
            text: this.operation?.getDescription(),
          },
        ],
      },
    ];
  }

  generatorParamsSerializer(): string {
    return `paramsSerializer(params:${this.getParametersName(
      "type",
      "queryRequestName",
    )}) {
            return qs.stringify(params)
        }`;
  }

  /**
   * @example
   * ```
   * request({
   *  method:"get",
   *  headers:{
   *    'Content-type':'application/x-www-form-urlencoded'
   *  },
   *  url:"/xxx/xx",
   *  params:query,
   *  data:body,
   * paramsSerializer(params: BusinessType.DetailByIdQueryRequest) {
   *                 return qs.stringify(params)
   *             }
   * })
   * ```
   */
  generatorMethodBody(): string {
    const url = new URLPath(<string>this.operation?.path);
    // headers: { 'Content-Type': 'multipart/form-data' },
    const header = () => {
      if (this.openapi.requestBody?.isJsonContainsDefaultCases) {
        return "";
      }
      return `headers:{
        'Content-Type':'${this.openapi.requestBody?.getContentType}'
    }`;
    };

    const requestFuncContent = _.chain([] as string[])
      .push("method:" + "'" + this.operation?.method + "'")
      .push(header())
      .push("url:" + url.requestPath)
      .push(this.openapi.parameter?.hasQueryParameters ? "params:query" : "")
      .push(this.openapi.requestBody?.hasRequestBody ? "data:body" : "")
      .push(this.openapi.response?.isDownLoad ? "responseType:'blob'" : "")
      .push(
        this.openapi.parameter?.hasQueryParametersArray
          ? this.generatorParamsSerializer()
          : "",
      )
      .filter(Boolean)
      .join(",\n")
      .value();
    return `return request({
             ${requestFuncContent}
})`;
  }

  generatorMethod(): MethodDeclarationStructure {
    const statement = {
      name: this.openapi.requestName,
      decorators: this.generatorMethodDecorators(),
      parameters: this.generatorMethodParameters(),
      returnType: this.generatorReturnType(),
      docs: this.generatorMethodDocs(),
      // scope: Scope.Public,
      statements: this.generatorMethodBody(),
    };
    return this.ast.generateMethodStatements(statement);
  }
}
