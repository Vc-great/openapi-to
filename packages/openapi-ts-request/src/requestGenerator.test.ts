import path from "node:path";
import { dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { AST, OpenAPI } from "@openapi-to/core";

import _ from "lodash";
import Oas from "oas";
import { describe } from "vitest";

import petStore from "../mock/petstore.json";
import { RequestGenerator } from "./RequestGenerator";

import type { OpenapiToSingleConfig } from "@openapi-to/core";
import type { PluginConfig } from "./types.ts";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe("RequestGenerator", async () => {
  const openapiToSingleConfig: OpenapiToSingleConfig = {
    input: {
      path: "",
      name: "",
    },
    output: {
      dir: path.resolve(__dirname, "../mock"),
    },
    plugins: [],
  };

  const context = {
    output: "",
  };

  test("createZodDecorator true getFullText", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);
    const pluginConfig = {
      createZodDecorator: true,
      compare: false,
    };

    const requestGenerator = new RequestGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    requestGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });

  test("compare createZodDecorator true getFullText", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);
    const pluginConfig = {
      createZodDecorator: true,
      compare: true,
    };

    const requestGenerator = new RequestGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    requestGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });

  test("createZodDecorator false getFullText", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);
    const pluginConfig = {
      createZodDecorator: false,
    };

    const requestGenerator = new RequestGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    requestGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });

  test("compare createZodDecorator false getFullText", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);
    const pluginConfig = {
      createZodDecorator: false,
      compare: true,
    };

    const requestGenerator = new RequestGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    requestGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });

  test("zodDecoratorImportDeclaration requestImportDeclaration  getFullText", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);
    const pluginConfig: PluginConfig = {
      createZodDecorator: true,
      compare: false,
      zodDecoratorImportDeclaration: {
        moduleSpecifier: "./test/zod",
      },
      requestImportDeclaration: {
        moduleSpecifier: "./test/request",
      },
    };

    const requestGenerator = new RequestGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    requestGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });
});
