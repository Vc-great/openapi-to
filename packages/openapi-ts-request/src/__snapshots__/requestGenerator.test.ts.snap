// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RequestGenerator > requestGenerator getFullText 1`] = `
"import { PetAPIZod } from "./PetAPIZod";
import { paramsZodSchema, responseZodSchema, zodValidate } from "@/utils/zod";
import { PetAPIType } from "./PetAPIType";
import { ErrorResponse } from "./model/ErrorResponse";
import { request } from "./request";

/**
 *
 * @tag pet
 * @description Everything about your Pets
 */
class PetAPI {
    /**
     *
     * @summary summary
     * @description
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.TestPostResponse)
    testPost(@paramsZodSchema(PetAPIZod.TestPostBodyRequest) body: PetAPIType.TestPostBodyRequest): Promise<[ErrorResponse, PetAPIType.TestPostResponse]> {
        return request({
            method: 'post',
            url: \`/pet/test\`,
            data: body
        })
    }

    /**
     *
     * @summary summary
     * @description
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.TestPutResponse)
    testPut(@paramsZodSchema(PetAPIZod.TestPutBodyRequest) body: PetAPIType.TestPutBodyRequest): Promise<[ErrorResponse, PetAPIType.TestPutResponse]> {
        return request({
            method: 'put',
            url: \`/pet/test\`,
            data: body
        })
    }

    /**
     *
     * @summary summary
     * @description
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.DelByTestResponse)
    delByTest(@paramsZodSchema(PetAPIZod.DelByTestBodyRequest) body: PetAPIType.DelByTestBodyRequest): Promise<[ErrorResponse, PetAPIType.DelByTestResponse]> {
        return request({
            method: 'delete',
            url: \`/pet/test\`,
            data: body
        })
    }

    /**
     *
     * @summary summary
     * @description
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.TestIdResponse)
    testId(@paramsZodSchema(PetAPIZod.TestIdQueryRequest) query: PetAPIType.TestIdQueryRequest, @paramsZodSchema(PetAPIZod.TestIdPathRequest) testId: PetAPIType.TestIdPathRequest): Promise<[ErrorResponse, PetAPIType.TestIdResponse]> {
        return request({
            method: 'get',
            url: \`/pet/test/\${testId}\`,
            params: query,
            paramsSerializer(params: PetAPIType.TestIdQueryRequest) {
                return qs.stringify(params)
            }
        })
    }

    /**
     *
     * @summary uploads an image
     * @description pet
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.UploadImagePostResponse)
    uploadImagePost(@paramsZodSchema(PetAPIZod.UploadImagePostBodyRequest) body: PetAPIType.UploadImagePostBodyRequest, @paramsZodSchema(PetAPIZod.UploadImagePostPathRequest) petId: PetAPIType.UploadImagePostPathRequest): Promise<[ErrorResponse, PetAPIType.UploadImagePostResponse]> {
        return request({
            method: 'post',
            headers: {
                'Content-Type': 'multipart/form-data'
            },
            url: \`/pet/\${petId}/uploadImage\`,
            data: body
        })
    }

    /**
     *
     * @summary Add a new pet to the store
     * @description
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.CreateResponse)
    create(@paramsZodSchema(PetAPIZod.CreateBodyRequest) body: PetAPIType.CreateBodyRequest): Promise<[ErrorResponse, PetAPIType.CreateResponse]> {
        return request({
            method: 'post',
            url: \`/pet\`,
            data: body
        })
    }

    /**
     *
     * @summary Update an existing pet
     * @description
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.UpdateResponse)
    update(@paramsZodSchema(PetAPIZod.UpdateBodyRequest) body: PetAPIType.UpdateBodyRequest): Promise<[ErrorResponse, PetAPIType.UpdateResponse]> {
        return request({
            method: 'put',
            url: \`/pet\`,
            data: body
        })
    }

    /**
     *
     * @summary Finds Pets by status
     * @description Multiple status values can be provided with comma separated strings
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.FindByStatusGetResponse)
    findByStatusGet(@paramsZodSchema(PetAPIZod.FindByStatusGetQueryRequest) query: PetAPIType.FindByStatusGetQueryRequest): Promise<[ErrorResponse, PetAPIType.FindByStatusGetResponse]> {
        return request({
            method: 'get',
            url: \`/pet/findByStatus\`,
            params: query,
            paramsSerializer(params: PetAPIType.FindByStatusGetQueryRequest) {
                return qs.stringify(params)
            }
        })
    }

    /**
     *
     * @summary Finds Pets by tags
     * @description Multiple tags can be provided with comma separated strings. Use tag1, tag2, tag3 for testing.
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.FindByTagsGetResponse)
    findByTagsGet(@paramsZodSchema(PetAPIZod.FindByTagsGetQueryRequest) query: PetAPIType.FindByTagsGetQueryRequest): Promise<[ErrorResponse, PetAPIType.FindByTagsGetResponse]> {
        return request({
            method: 'get',
            url: \`/pet/findByTags\`,
            params: query,
            paramsSerializer(params: PetAPIType.FindByTagsGetQueryRequest) {
                return qs.stringify(params)
            }
        })
    }

    /**
     *
     * @summary Find pet by ID
     * @description Returns a single pet
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.FindByPetIdResponse)
    findByPetId(@paramsZodSchema(PetAPIZod.FindByPetIdPathRequest) petId: PetAPIType.FindByPetIdPathRequest): Promise<[ErrorResponse, PetAPIType.FindByPetIdResponse]> {
        return request({
            method: 'get',
            url: \`/pet/\${petId}\`
        })
    }

    /**
     *
     * @summary Updates a pet in the store with form data
     * @description
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.PetIdResponse)
    petId(@paramsZodSchema(PetAPIZod.PetIdBodyRequest) body: PetAPIType.PetIdBodyRequest, @paramsZodSchema(PetAPIZod.PetIdPathRequest) petId: PetAPIType.PetIdPathRequest): Promise<[ErrorResponse, PetAPIType.PetIdResponse]> {
        return request({
            method: 'post',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            url: \`/pet/\${petId}\`,
            data: body
        })
    }

    /**
     *
     * @summary Deletes a pet
     * @description
     */
    @zodValidate
    @responseZodSchema(PetAPIZod.DelByPetIdResponse)
    delByPetId(@paramsZodSchema(PetAPIZod.DelByPetIdPathRequest) petId: PetAPIType.DelByPetIdPathRequest): Promise<[ErrorResponse, PetAPIType.DelByPetIdResponse]> {
        return request({
            method: 'delete',
            url: \`/pet/\${petId}\`
        })
    }
}

export const petAPI = new PetAPI;

import { StoreAPIZod } from "./StoreAPIZod";
import { paramsZodSchema, responseZodSchema, zodValidate } from "@/utils/zod";
import { StoreAPIType } from "./StoreAPIType";
import { ErrorResponse } from "./model/ErrorResponse";
import { request } from "./request";

/**
 *
 * @tag store
 * @description Access to Petstore orders
 */
class StoreAPI {
    /**
     *
     * @summary Place an order for a pet
     * @description
     */
    @zodValidate
    @responseZodSchema(StoreAPIZod.CreateResponse)
    create(@paramsZodSchema(StoreAPIZod.CreateBodyRequest) body: StoreAPIType.CreateBodyRequest): Promise<[ErrorResponse, StoreAPIType.CreateResponse]> {
        return request({
            method: 'post',
            url: \`/store/order\`,
            data: body
        })
    }

    /**
     *
     * @summary Find purchase order by ID
     * @description For valid response try integer IDs with value >= 1 and <= 10. Other values will generated exceptions
     */
    @zodValidate
    @responseZodSchema(StoreAPIZod.FindByOrderIdResponse)
    findByOrderId(@paramsZodSchema(StoreAPIZod.FindByOrderIdPathRequest) orderId: StoreAPIType.FindByOrderIdPathRequest): Promise<[ErrorResponse, StoreAPIType.FindByOrderIdResponse]> {
        return request({
            method: 'get',
            url: \`/store/order/\${orderId}\`
        })
    }

    /**
     *
     * @summary Delete purchase order by ID
     * @description For valid response try integer IDs with positive integer value. Negative or non-integer values will generate API errors
     */
    @zodValidate
    @responseZodSchema(StoreAPIZod.DelByOrderIdResponse)
    delByOrderId(@paramsZodSchema(StoreAPIZod.DelByOrderIdPathRequest) orderId: StoreAPIType.DelByOrderIdPathRequest): Promise<[ErrorResponse, StoreAPIType.DelByOrderIdResponse]> {
        return request({
            method: 'delete',
            url: \`/store/order/\${orderId}\`
        })
    }

    /**
     *
     * @summary Returns pet inventories by status
     * @description Returns a map of status codes to quantities
     */
    @zodValidate
    @responseZodSchema(StoreAPIZod.InventoryGetResponse)
    inventoryGet(): Promise<[ErrorResponse, StoreAPIType.InventoryGetResponse]> {
        return request({
            method: 'get',
            url: \`/store/inventory\`
        })
    }
}

export const storeAPI = new StoreAPI;

import { UserAPIZod } from "./UserAPIZod";
import { paramsZodSchema, responseZodSchema, zodValidate } from "@/utils/zod";
import { UserAPIType } from "./UserAPIType";
import { ErrorResponse } from "./model/ErrorResponse";
import { request } from "./request";

/**
 *
 * @tag user
 * @description Operations about user
 */
class UserAPI {
    /**
     *
     * @summary Creates list of users with given input array
     * @description
     */
    @zodValidate
    @responseZodSchema(UserAPIZod.CreateWithArrayPostResponse)
    createWithArrayPost(@paramsZodSchema(UserAPIZod.CreateWithArrayPostBodyRequest) body: UserAPIType.CreateWithArrayPostBodyRequest): Promise<[ErrorResponse, UserAPIType.CreateWithArrayPostResponse]> {
        return request({
            method: 'post',
            url: \`/user/createWithArray\`,
            data: body
        })
    }

    /**
     *
     * @summary Creates list of users with given input array
     * @description
     */
    @zodValidate
    @responseZodSchema(UserAPIZod.CreateWithListPostResponse)
    createWithListPost(@paramsZodSchema(UserAPIZod.CreateWithListPostBodyRequest) body: UserAPIType.CreateWithListPostBodyRequest): Promise<[ErrorResponse, UserAPIType.CreateWithListPostResponse]> {
        return request({
            method: 'post',
            url: \`/user/createWithList\`,
            data: body
        })
    }

    /**
     *
     * @summary Get user by user name
     * @description
     */
    @zodValidate
    @responseZodSchema(UserAPIZod.FindByUsernameResponse)
    findByUsername(@paramsZodSchema(UserAPIZod.FindByUsernamePathRequest) username: UserAPIType.FindByUsernamePathRequest): Promise<[ErrorResponse, UserAPIType.FindByUsernameResponse]> {
        return request({
            method: 'get',
            url: \`/user/\${username}\`
        })
    }

    /**
     *
     * @summary Updated user
     * @description This can only be done by the logged in user.
     */
    @zodValidate
    @responseZodSchema(UserAPIZod.UsernameResponse)
    username(@paramsZodSchema(UserAPIZod.UsernameBodyRequest) body: UserAPIType.UsernameBodyRequest, @paramsZodSchema(UserAPIZod.UsernamePathRequest) username: UserAPIType.UsernamePathRequest): Promise<[ErrorResponse, UserAPIType.UsernameResponse]> {
        return request({
            method: 'put',
            url: \`/user/\${username}\`,
            data: body
        })
    }

    /**
     *
     * @summary Delete user
     * @description This can only be done by the logged in user.
     */
    @zodValidate
    @responseZodSchema(UserAPIZod.DelByUsernameResponse)
    delByUsername(@paramsZodSchema(UserAPIZod.DelByUsernamePathRequest) username: UserAPIType.DelByUsernamePathRequest): Promise<[ErrorResponse, UserAPIType.DelByUsernameResponse]> {
        return request({
            method: 'delete',
            url: \`/user/\${username}\`
        })
    }

    /**
     *
     * @summary Logs user into the system
     * @description
     */
    @zodValidate
    @responseZodSchema(UserAPIZod.LoginGetResponse)
    loginGet(@paramsZodSchema(UserAPIZod.LoginGetQueryRequest) query: UserAPIType.LoginGetQueryRequest): Promise<[ErrorResponse, UserAPIType.LoginGetResponse]> {
        return request({
            method: 'get',
            url: \`/user/login\`,
            params: query
        })
    }

    /**
     *
     * @summary Logs out current logged in user session
     * @description
     */
    @zodValidate
    @responseZodSchema(UserAPIZod.LogoutGetResponse)
    logoutGet(): Promise<[ErrorResponse, UserAPIType.LogoutGetResponse]> {
        return request({
            method: 'get',
            url: \`/user/logout\`
        })
    }

    /**
     *
     * @summary Create user
     * @description This can only be done by the logged in user.
     */
    @zodValidate
    @responseZodSchema(UserAPIZod.CreateResponse)
    create(@paramsZodSchema(UserAPIZod.CreateBodyRequest) body: UserAPIType.CreateBodyRequest): Promise<[ErrorResponse, UserAPIType.CreateResponse]> {
        return request({
            method: 'post',
            url: \`/user\`,
            data: body
        })
    }
}

export const userAPI = new UserAPI;
"
`;
